From b88b6f6f3a29b70cd0b314da8ceab18b0b34eed6 Mon Sep 17 00:00:00 2001
From: Jeremy Harris <jgh146exb@wizmail.org>
Date: Sat, 9 Feb 2019 16:56:59 +0000
Subject: [PATCH 01/29] Fix json extract operator for unfound case

(cherry picked from commit e73798976812e652320f096870359ef35ed069ff)
(cherry picked from commit b2734f7b45111f9b7de790c7b334a2ece47675b5)
---
 doc/doc-docbook/spec.xfpt    | 4 ++++
 src/expand.c             | 9 +++++----
 test/scripts/0000-Basic/0002 | 3 +++
 test/stdout/0002             | 3 +++
 4 files changed, 15 insertions(+), 4 deletions(-)

--- a/src/expand.c
+++ b/src/expand.c
@@ -5861,10 +5861,11 @@ while (*s != 0)
 	      }
 	    while (field_number > 0 && (item = json_nextinlist(&list)))
 	      field_number--;
-	    s = item;
-	    lookup_value = s;
-	    while (*s) s++;
-	    while (--s >= lookup_value && isspace(*s)) *s = '\0';
+	    if ((lookup_value = s = item))
+	      {
+	      while (*s) s++;
+	      while (--s >= lookup_value && isspace(*s)) *s = '\0';
+	      }
 	    }
 	  else
 	    {
--- a/doc/spec.txt
+++ b/doc/spec.txt
@@ -8776,6 +8776,8 @@ ${extract json{<key>}{<string1>}{<string
     The braces, commas and colons, and the quoting of the member name are
     required; the spaces are optional. Matching of the key against the member
     names is done case-sensitively.
+    If a returned value is a JSON string, it retains its leading and
+    trailing quotes.
 
     The results of matching are handled as above.
 
@@ -8813,6 +8815,8 @@ ${extract json{<number>}}{<string1>}{<st
 
     Field selection and result handling is as above; there is no choice of
     field separator.
+    If a returned value is a JSON string, it retains its leading and
+    trailing quotes.
 
 ${filter{<string>}{<condition>}}
 
